src/Root.ts

```diff
@@ -203,54 +200,42 @@ export class Root
         if (element) {
             if (hasFocused) {
                 this._isFocused = true;
+                this._dummyManager?.setTabbable(false);
+
                 triggerEvent<Types.RootFocusEventDetails>(
                     this._tabster.root.eventTarget,
                     "focus",
-                    { element, fromAdjacent }
+                    { element }
                 );
             } else {
                 this._setFocusedTimer = this._tabster
                     .getWindow()
                     .setTimeout(() => {
                         delete this._setFocusedTimer;
+
                         this._isFocused = false;
+                        this._dummyManager?.setTabbable(true);
+
                         triggerEvent<Types.RootFocusEventDetails>(
                             this._tabster.root.eventTarget,
                             "blur",
-                            { element, fromAdjacent }
+                            { element }
                         );
                     }, 0);
             }
         }
     };
 
-    private _onFocus = (e: HTMLElement | undefined) => {
-        const win = this._tabster.getWindow();
-
-        if (this._setTabbableTimer) {
-            win.clearTimeout(this._setTabbableTimer);
-            delete this._setTabbableTimer;
-        }
-
-        if (e) {
-            const ctx = RootAPI.getTabsterContext(this._tabster, e);
-
-            if (ctx) {
-                this._setFocused(ctx.root.getElement() === this._element.get());
-            }
+    private _onFocusIn = (event: FocusEvent) => {
+        const target = event.target as HTMLElement | null;
 
-            if (!ctx || ctx.uncontrolled || this._tabster.rootDummyInputs) {
-                this._dummyManager?.setTabbable(false);
-                return;
-            }
-        } else {
-            this._setFocused(false);
+        if (target && this._element.get()?.contains(target)) {
```

Comment:
It's possible that the root is a virtual parent of the focused element and not a direct DOM parent. Do you know if that that would cause problems if it doesn't uses the changes in https://github.com/microsoft/tabster/pull/308?

Reply:
Added getParent(). And I want to go through the virtual parent scenarios deeper next. To create more tests as I suspect we might have corner cases.